personalrobotics/ada_feeding
# Implement Face Detection

# Description

This PR adds a face detection node, addressing https://github.com/personalrobotics/ada_feeding/issues/27. It also addresses some issues with segment_from_point (python 3 typing errors and a function call that was possibly blocking realsense nodes).

# Testing procedure

On the realsense camera (using t0b1/nano), I tested that the node starts and stops with the appropriate service call, and publishes mouth locations only when a face is detected.

**Initialization**:
1. SSH into nano@nano
2. On nano, start vision nodes:
    1. `ros2config`
    2. `ros2 launch realsense2_camera rs_launch.py rgb_camera.profile:='640,480,30' depth_module.profile:='640,480,30' align_depth.enable:='true' initial_reset:='true'`
3. On t0b1, launch the face_detection node:
    1. `ros2 launch ada_feeding_perception ada_feeding_perception_launch.xml`
4. Listen to the /face_detection topic:
    1. `ros2 topic echo /face_detection`

**Testing**:
1. For each of the following commands, run it and verify that it executes as expected. 
    1. Turn on face detection 
        - [x] `ros2 service call /toggle_face_detection std_srvs/srv/SetBool "{data: True}"`
        - [x] The face detection node should receive this message
   2. Turn off face detection
        - [x]  `ros2 service call /toggle_face_detection std_srvs/srv/SetBool "{data: False}"`
        - [x] The face detection node should receive this message
2. Position your face in front of the camera and move to several different positions. Make sure that the face_detection topic returns the correct values.
     

# Before opening a pull request

# Before Merging
